File: 70-climbing-stairs/climbing-stairs.py
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        
        """
        there are only two ways to climb to the top:

        1. 1 step
        2. 2 step

        basically
        """
        memo = {}
        def dp(num):
            if num in memo:
                return memo[num]

            if num > n:
                return 0

            if num == n:
                return 1

            memo[num] = dp(num + 1) + dp(num+2)  
            return memo[num]
            
        return dp(0)

File: 70-climbing-stairs/climbing-stairs.cpp
class Solution {
public:
    int DP(int n, unordered_map<int,int>& memo) {
        // base case
        if (n == 0 || n == 1) {
            return 1;
        }

        // check for precomputed solution
        if (memo.find(n) == memo.end()) {
            memo[n] = DP(n-1, memo) + DP(n-2, memo);
        }

        return memo[n];
    }

    int climbStairs(int n) {
        unordered_map<int, int> memo;
        return DP(n, memo);
    }
};